keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1535 @@
+be able to identify other evacuators and an order in which they will run.
+
+#### Story 4
+
+As an application owner I want my pods to be scheduled on correct nodes. I want to use the
+descheduler or the upcoming Affinity Based Eviction feature to remove pods from incorrect nodes
+and then have the pods scheduled on new ones. I want to do the rescheduling gracefully and be able
+to control the disruption level of my application (even 0% application unavailability).
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Length Limitations for Pod Annotations and Evacuation Finalizers
+
+We use the following finalizers in Evacuation objects:
+`evacuation.coordination.k8s.io/instigator_${EVACUATION_INSTIGATOR_SUBDOMAIN}`.
+
+We use the following annotations in Pod objects:
+`evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}: ${PRIORITY}/${ROLE}`.
+
+Key validation for both of these is subject to [IsQualifiedName](https://github.com/kubernetes/kubernetes/blob/cae35dba5a3060711a2a3f958537003bc74a59c0/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L42)
+validation. The key consists of `prefix/name`. The maximum length of the name is 63 characters,
+which may not always be enough for `EVACUATION_INSTIGATOR_SUBDOMAIN` and `EVACUATOR_CLASS` (also
+DNS subdomain) after adding the length of `instigator_` and `priority_` respectively.
+
+If the name component of the key is greater than 63 characters, it is recommended to use a unique
+shortcut of your subdomain name.
+For example, `deployment.apps.k8s.io` instead of `deployment.apps.kubernetes.io`.
+
+We do not expect a large number of instigators and evacuators. If this becomes a larger problem in
+the future, we can implement an additional scheme:
+
+```yaml
+annotations:
+  evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS_HASH}: ${PRIORITY}/${ROLE}
+  evacuation.coordination.k8s.io/hash_${EVACUATOR_CLASS_HASH}: ${EVACUATOR_CLASS}
+```
+
+which would result in
+
+```yaml
+annotations:
+  evacuation.coordination.k8s.io/priority_7bb77f87dc: "11000/knowledgeable-app-specific"
+  evacuation.coordination.k8s.io/hash_7bb77f87dc: "sensitive-workload-operator.tom-and-jerry-delicious-fruit-company.com"
+```
+
+`7bb77f87dc` value is a hash of `sensitive-workload-operator.tom-and-jerry-delicious-fruit-company.com`
+
+### Risks and Mitigations
+
+If there is no evacuator, and the application has insufficient availability and a blocking PDB or
+blocking validating admission webhook, then the evacuation controller will enter into an eviction
+cold loop. To mitigate this we will increment `.status.failedEvictionCounter` and `X` metric.
+
+An evacuator could reconcile the status properly without making any progress. It is thus
+recommended to check `creationTimestamp` of the Evacuations and observe
+`evacuation_controller_active_evacuator` metric to see how much it takes for an evacuator to
+complete the evacuation. This metric can be also used to implement additional alerts.
```

Comment:
What about the risk behind adopting this new API, most of the users are very slow adopting these. My concern is that, aside from the naming confusion overload as mentioned in https://github.com/kubernetes/enhancements/pull/4565/files#r1543242249, a lot of users will prefer the simpler eviction API that is already battle tested even with its limitation, but mostly because it's already being widely used. Your proposed mechanism on one hand requires work on the k8s core side, and on the other hand to application owners managing the evacuation process. 

I will repeat similarly to what has been stated in https://github.com/kubernetes/enhancements/pull/4213, should we revisit PDB and eviction API, both of which are widely used and identify the gaps and what are the options to extend them with the missing functionality. I don't think we need a full fledged, feature-rich functionality, but rather a functional and easy to understand API built on the existing primitives.

Reply:
I do not see a big risk in terms of adoption.

The eviction API is mostly used by a handful of components (kubectl drain, cluster autoscaler, karpenter, scheduler, descheduler, etc.). It is enough to introduce Evacuation API  to a handful of components to see an impact. Users usually do not use this API by themselves - there is not even a kubectl support.

From the evacuator side, there is no user action required either. Please see the followup workload integration in this KEP, DeamonSet is part of the NodeMaintenance KEP.

The user can take action if they want to further improve the situation:
- Implement custom evacuation strategy for their workload or use a strategy implemented by another party/component.
- Add additional checks to the pods (aspect oriented approach) as discussed in https://github.com/kubernetes/enhancements/pull/4213#discussion_r1537878392.


At this time, I do not see any extensions that could be made to PDBs and eviction API to solve the current issues. Please see the analysis in the alternatives section: https://github.com/atiratree/kube-enhancements/blob/improve-node-maintenance/keps/sig-apps/4212-declarative-node-maintenance/README.md#alternatives.

Reply:
Apart from the NodeMaintenance we have already have two additional use cases that would like to adopt Evacuation API
 
- Affinity Based Eviction (https://github.com/kubernetes/enhancements/issues/4328)
- Descheduler integration (https://github.com/kubernetes-sigs/descheduler/pull/1354)

Reply:
> Descheduler integration (https://github.com/kubernetes-sigs/descheduler/pull/1354)
Why not align with the effort that scheduler is pushing to start using evictions (see https://github.com/kubernetes/enhancements/issues/3280)?

Reply:
That is too specific for preemption and expects ordered pods by a priority. It would not help in cases where priority is not used or cooperative eviction is needed.